compositor_render/src/renderer/texture.rs

```diff
@@ -32,4 +38,225 @@ impl Texture {
+        ctx.queue.submit([]);
+    }
+
+    fn upload_data(&self, queue: &wgpu::Queue, data: &[u8], bytes_per_pixel: u32) {
+        queue.write_texture(
+            wgpu::ImageCopyTexture {
+                aspect: wgpu::TextureAspect::All,
+                mip_level: 0,
+                origin: wgpu::Origin3d::ZERO,
+                texture: &self.texture,
+            },
+            data,
+            wgpu::ImageDataLayout {
+                offset: 0,
+                bytes_per_row: Some(self.texture.width() * bytes_per_pixel),
+                rows_per_image: Some(self.texture.height()),
+            },
+            self.texture.size(),
+        );
+    }
+}
+
+pub struct OutputTexture {
+    yuv_textures: [Texture; 3],
+    buffers: [wgpu::Buffer; 3],
+    pub resolution: Resolution,
+}
+
+impl OutputTexture {
+    fn padded(width: usize) -> usize {
+        width + (256 - (width % 256))
+    }
+    pub fn new(ctx: &WgpuCtx, resolution: &Resolution) -> Self {
+        let device = &ctx.device;
+        let buffers = [
+            device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some("output texture buffer 0"),
+                mapped_at_creation: false,
+                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
+                size: (Self::padded(resolution.width) * resolution.height) as u64,
+            }),
+            device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some("output texture buffer 1"),
+                mapped_at_creation: false,
+                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
+                size: (Self::padded(resolution.width / 2) * resolution.height / 2) as u64,
+            }),
+            device.create_buffer(&wgpu::BufferDescriptor {
+                label: Some("output texture buffer 2"),
+                mapped_at_creation: false,
+                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
+                size: (Self::padded(resolution.width / 2) * resolution.height / 2) as u64,
+            }),
+        ];
+        Self {
+            yuv_textures: Texture::new_yuv_textures(ctx, resolution),
+            buffers,
+            resolution: *resolution,
+        }
+    }
+
+    /// extremely hacky workaround to download texture
+    /// TODO: do it properly
+    pub fn download(&self, ctx: &WgpuCtx) -> YuvData {
+        self.transfer_content_to_buffers(&ctx.device, &ctx.queue);
+        let mut result = YuvData {
+            y_plane: Bytes::new(),
+            u_plane: Bytes::new(),
+            v_plane: Bytes::new(),
+        };
+        {
+            let size = self.yuv_textures[0].texture.size();
+            let buffer = BytesMut::with_capacity(size.width as usize * size.height as usize);
+
+            self.buffers[0]
+                .slice(..)
+                .map_async(MapMode::Read, |x| x.unwrap());
+            ctx.device.poll(wgpu::MaintainBase::Wait);
```

Comment:
If you want to make this code simpler and nicer in this PR, I found a potentially better way of doing this using [`wgpu::util::DownloadBuffer`](https://docs.rs/wgpu/latest/wgpu/util/struct.DownloadBuffer.html). Otherwise you could mention the `DownloadBuffer` in the `TODO`.

Reply:
I'll leave it for follow-up PR, but thank you for the link.